README.md
m4atx_battery_monitor [![Build Status](https://api.travis-ci.org/WPI-RAIL/m4atx_battery_monitor.png)](https://travis-ci.org/WPI-RAIL/m4atx_battery_monitor)
=====================

#### Battery Monitor for the M4-ATX Power Module
For full documentation, see [the ROS wiki](http://ros.org/wiki/m4atx_battery_monitor).

The node reads in and publishes the status of the battery supply. It also verbalizes the battery low warnings using `espeak`.

### Setup
`$ lsusb`

Find Microchip Technology, Inc. and mark down the bus and device numbers.

`$ sudo chmod a+rw /dev/bus/usb/[bus #]/[device #]`

Now the node will connect and run properly.

### License
For full terms and conditions, see the [LICENSE](LICENSE) file.

### Authors
See the [AUTHORS](AUTHORS.md) file for a full list of contributors.


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package m4atx_battery_monitor
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.2 (2014-08-05)
------------------
* moved .sh to .bash
* permissions script added to install
* Contributors: Russell Toris

* usb dev added
* comments
* moved script
* Merge branch 'cmdunkers-develop' into develop
* merege
* Added a script to make sure the device can be read from
  script applies appropriate permissions
* revert to 0.0.0'
* Added a startup script
* Contributors: Chris Dunkers, Russell Toris

0.0.1 (2014-07-23)
------------------
* changelog updated
* espeak now used
* node renamed
* cleanup
* Merge pull request #1 from cmdunkers/develop
  battery monitor for the car battery
* added header to fix error
* attemp to fix broken pipe by ensuring usb closed after exiting code
* Updated the conversion for the Voltage input to more accurately represent the input voltage
* corrected the include file
* changed the m4api to the newer updated version/ the more accurate one
* added in sound debug
* added a line to print the iginition voltage, should be 0
* Added smicolons
* added usb reset code
* Added a values to the message so the value can be linked to an decription
  Also cleaned up the node
* cleared the build folder
* Merge branch 'develop' of https://github.com/cmdunkers/m4atx_battery_monitor into develop
* merged in changes
* reverted to previous ros node
  config
* testing out new node code
* added git ignore
* Deleted the build folder and added a git ignore
* Got the other api to make
* added source code from umd
* minor fixups
* Wrote the code but not yet tested
* Initial commit
* Contributors: Chris Dunkers, Christopher Dunkers, Russell Toris


src/m4atx_battery_monitor.cpp
/*!
 * M4-ATX diagnostics and configuration node.
 *
 * LGPL (c) 2010 Ken Tossell <ktossell@umd.edu>
 *
 * \author Ken Tossell <ktossell@umd.edu>
 * \author Chris Dunkers, WPI - cmdunkers@wpi.edu
 * \author Russell Toris, WPI - rctoris@wpi.edu
 */

#include <ros/ros.h>
#include <m4atx_battery_monitor/PowerReading.h>
#include <usb.h>

extern "C"
{
#include "m4atx_battery_monitor/m4api.h"
}

/*!
 * Creates and runs the m4atx_battery_monitor node.
 *
 * \param argc argument count that is passed to ros::init
 * \param argv arguments that are passed to ros::init
 * \return EXIT_SUCCESS if the node runs correctly, or EXIT_FAILURE if an error occurs.
 */
int main(int argc, char **argv)
{
  // initialize the node
  ros::init(argc, argv, "m4atx_battery_monitor");

  // main node handle
  ros::NodeHandle node;
  ros::Time next_check = ros::Time::now();

  // attempt to initialize the USB device
  struct usb_dev_handle *dev = m4Init();
  if (!dev)
  {
    ROS_ERROR("Could not initialize the M4-ATX.");
    return EXIT_FAILURE;
  }

  // grab the parameters
  ros::NodeHandle private_nh("~");
  double diag_frequency;
  private_nh.param<double>("diag_frequency", diag_frequency, 1.0);
  double input_nominal;
  private_nh.param<double>("input_nominal", input_nominal, 13.8);
  double battery_dead_voltage;
  private_nh.param<double>("battery_dead_voltage", battery_dead_voltage, 10.5);

  // setup the publishers
  ros::Publisher diag_pub = node.advertise<m4atx_battery_monitor::PowerReading>("battery_status_m4atx", 1);

  // main publish loop
  ros::Rate loop_rate(diag_frequency);
  while (ros::ok())
  {
    m4Diagnostics diag;
    m4atx_battery_monitor::PowerReading reading;

    // attempt to red the battery
    if (m4GetDiag(dev, &diag))
    {
      ROS_ERROR("Error reading from the M4-ATX.");
      return EXIT_FAILURE;
    }

    // parse the feedback
    reading.volts_read_item.push_back("VIN");
    reading.volts_read_value.push_back(diag.vin);
    reading.volts_read_item.push_back("12V");
    reading.volts_read_value.push_back(diag.v12);
    reading.volts_read_item.push_back("5V");
    reading.volts_read_value.push_back(diag.v5);
    reading.volts_read_item.push_back("3.3V");
    reading.volts_read_value.push_back(diag.v33);
    reading.volts_read_item.push_back("VIGN");
    reading.volts_read_value.push_back(diag.vign);
    reading.volts_full_item.push_back("VIN");
    reading.volts_full_value.push_back(input_nominal);
    reading.volts_full_item.push_back("12V");
    reading.volts_full_value.push_back(12.0);
    reading.volts_full_item.push_back("5V");
    reading.volts_full_value.push_back(5.0);
    reading.volts_full_item.push_back("3.3V");
    reading.volts_full_value.push_back(3.3);
    reading.temperature.push_back(diag.temp);

    // calculate the soc value
    double soc = (diag.vin - battery_dead_voltage) / (input_nominal - battery_dead_voltage);
    soc *= 100;
    reading.input_soc = soc;
    reading.header.stamp = ros::Time::now();

    // publish the feedback
    diag_pub.publish(reading);

    // check if we need to say something to the user
    if (ros::Time::now() > next_check)
    {
      if (soc < 5)
      {
        next_check = ros::Time::now() + ros::Duration(5.0 * 60); //5 min
        system("espeak \"Computer battery has 5 percent remaining.\"");
      }
      else if (soc < 10)
      {
        next_check = ros::Time::now() + ros::Duration(10.0 * 60); //10 min
        system("espeak \"Computer battery has 10 percent remaining.\"");
      }
      else if (soc < 20)
      {
        next_check = ros::Time::now() + ros::Duration(15.0 * 60); //15 min
        system("espeak \"Computer battery has 20 percent remaining.\"");
      }
      else
      {
        next_check = ros::Time::now() + ros::Duration(20.0 * 60); //20 min
      }
    }

    loop_rate.sleep();
  }

  //close the USB handle
  usb_close(dev);

  return EXIT_SUCCESS;
}


src/m4atx_battery_monitor_node.cpp
/* M4-ATX diagnostics and configuration node
 *
 * LGPL (c) 2010 Ken Tossell <ktossell@umd.edu>
 */

#include "ros/ros.h"
#include "m4atx_battery_monitor/PowerReading.h"
#include "std_msgs/String.h"
#include <usb.h>

extern "C" {
  #include "m4atx_battery_monitor/m4api.h"
}

int main (int argc, char **argv) 
{
	ros::init(argc, argv, "m4atx");

	ros::NodeHandle node;
	ros::Publisher diag_pub, sound_pub;
	double diag_frequency;
	double input_nominal;
	double battery_dead_voltage;
	ros::Time next_check = ros::Time::now();
	
	struct usb_dev_handle *dev;

	dev = m4Init();

	if (!dev) 
	{
		perror("Initializing M4-ATX");
		return -1;
	}

	ros::NodeHandle private_node_handle_("~");
	private_node_handle_.param<double>("diag_frequency", diag_frequency, 1.0);
	private_node_handle_.param<double>("input_nominal", input_nominal, 13.8);
	private_node_handle_.param<double>("battery_dead_voltage", battery_dead_voltage, 10.5);

	diag_pub = node.advertise<m4atx_battery_monitor::PowerReading>("battery_status_m4atx", 1);
	sound_pub = node.advertise<std_msgs::String>("say", 1);

	ros::Rate loop_rate(diag_frequency);
	while (ros::ok()) 
	{
		m4Diagnostics diag;
		m4atx_battery_monitor::PowerReading reading;

		if (m4GetDiag(dev, &diag)) 
		{
			perror("Reading from M4-ATX");
			return -1;
		}
		reading.volts_read_item.push_back("VIN");
		reading.volts_read_value.push_back(diag.vin);
		reading.volts_read_item.push_back("12V");
		reading.volts_read_value.push_back(diag.v12);
		reading.volts_read_item.push_back("5V");
		reading.volts_read_value.push_back(diag.v5);
		reading.volts_read_item.push_back("3.3V");
		reading.volts_read_value.push_back(diag.v33);
		reading.volts_read_item.push_back("VIGN");
		reading.volts_read_value.push_back(diag.vign);
		
		reading.volts_full_item.push_back("VIN");
		reading.volts_full_value.push_back(input_nominal);
		reading.volts_full_item.push_back("12V");
		reading.volts_full_value.push_back(12.0);
		reading.volts_full_item.push_back("5V");
		reading.volts_full_value.push_back(5.0);
		reading.volts_full_item.push_back("3.3V");
		reading.volts_full_value.push_back(3.3);

		reading.temperature.push_back(diag.temp);
		
		double soc = (diag.vin - battery_dead_voltage) / (input_nominal - battery_dead_voltage);
		soc *= 100;
		
		reading.input_soc = soc;
		
		reading.header.stamp = ros::Time::now();

		diag_pub.publish(reading);	
		
		if (ros::Time::now() > next_check)
		{
			std_msgs::String message; 
			if(soc < 5)
			{	
				next_check = ros::Time::now() + ros::Duration(5.0*60); //5 min 
				message.data = "Computer battery has 5 percent remaining";
				sound_pub.publish(message);
			}else if(soc < 10)
			{
				next_check = ros::Time::now() + ros::Duration(10.0*60); //10 min 
				message.data = "Computer battery has 10 percent remaining";
				sound_pub.publish(message);
			} else if(soc < 20)
			{
				next_check = ros::Time::now() + ros::Duration(15.0*60); //15 min
				message.data = "Computer battery has 20 percent remaining";
				sound_pub.publish(message);
			} else 
			{
				next_check = ros::Time::now() + ros::Duration(20.0*60); //20 min 	
			}
		}
		
		loop_rate.sleep();
	}
	
	//close the usb handle
	usb_close(dev);	
	
	return 0;
}



include/m4atx_battery_monitor/m4api.h
/*
 * m4api, a tool for M4-ATX DC-DC power supplies
 * (c) 2009-2010 Ken Tossell <ktossell@umd.edu>
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License, version 2.1,
 * as published by the Free Software Foundation. 
 *
 * This software is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this software; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#ifndef M4API_H
#define M4API_H

#ifdef __cplusplus
extern "C" {
#endif // __cplusplus

const int M4API_VERSION_MAJOR = 0;
const int M4API_VERSION_MINOR = 1;
const float M4API_VERSION = 0.1;

enum m4Type {
  M4_VLT_12_11,
  M4_VLT_12_07,
  M4_VLT_5_03,
  M4_VLT_33_01,
  M4_DEG,
  M4_SEC,
  M4_MSC_1_1, // 1 byte
  M4_MSC_1_2, // 2 bytes
  M4_MSC_10_1, // 1 byte
  M4_MSC_10_2, // 2 bytes
  M4_TIM, // 2 bytes
  M4_TRY,
  M4_BYT
};

enum m4FieldID {
  M4_IGN_HIGH,
  M4_IGN_LOW,
  M4_IGN_DBC,
  M4_PSU_DELAY,
  M4_VIN_MIN_START,
  M4_VIN_MIN_ON,
  M4_VIN_MIN_5V,
  M4_VIN_MAX,
  M4_12V_MAX,
  M4_12V_MIN,
  M4_5V_MAX,
  M4_5V_MIN,
  M4_33V_MAX,
  M4_33V_MIN,
  M4_12V_TIME,
  M4_33V_TIME,
  M4_PWRSW,
  M4_PSU_ON_TIME,
  M4_ON_DELAY,
  M4_PSU_OFF_TIME,
  M4_OFF_DELAY,
  M4_EMG_TIME_5VSB,
  M4_EMG_TIMER,
  M4_PS_ON_0,
  M4_PS_ON_1,
  M4_THUMP,
  M4_TEMP_MAX,
  M4_TEMP_MIN,
  M4_EMG_OFF_MODE,
  M4_5V_SBY_DLY,
  M4_OFF_DELAY_0,
  M4_OFF_HARD_0,
  M4_OFF_DELAY_1,
  M4_OFF_HARD_1,
  M4_OFF_DELAY_2,
  M4_OFF_HARD_2,
  M4_OFF_DELAY_3,
  M4_OFF_HARD_3,
  M4_OFF_DELAY_4,
  M4_OFF_HARD_4,
  M4_OFF_DELAY_5,
  M4_OFF_HARD_5,
  M4_OFF_DELAY_6,
  M4_OFF_HARD_6,
  M4_OFF_DELAY_7,
  M4_OFF_HARD_7,
  M4_RESET,
  M4_NUM_CONFIG_FIELDS
};

enum m4Repr {
  M4_INTEG,
  M4_FLOAT,
  M4_TIMER
};

struct m4DiagField {
  enum m4Type type;
  unsigned int index;
  char *name;
  char *desc;
};

struct m4ConfigField {
  enum m4Type type;
  unsigned int index;
  char *name;
  char *desc;
};

struct m4Diagnostics {
  float vin;
  float vign;
  float v33;
  float v5;
  float v12;
  float temp;
};

struct m4Version {  
  int major;
  int minor;
};

extern struct m4DiagField m4DiagFields[];
extern struct m4ConfigField m4ConfigFields[];
extern char* m4TypeDescs[];

/* Find and open the PSU */
struct usb_dev_handle *m4Init();

/* User-friendly-ish routines for setting/getting config values */
int m4GetFloat(struct usb_dev_handle *dev, enum m4FieldID fid, float *out);
int m4SetFloat(struct usb_dev_handle *dev, enum m4FieldID fid, float val);
int m4GetInteger(struct usb_dev_handle *dev, enum m4FieldID fid, int *out);
int m4SetInteger(struct usb_dev_handle *dev, enum m4FieldID fid, int val);

/* Get the diagnostic string in its raw form (<0 = error) */
int m4FetchDiag (struct usb_dev_handle *dev, char *buf);

/* Get the diagnostic values in processed form (<0 = error) */
int m4GetDiag (struct usb_dev_handle *dev, struct m4Diagnostics *diag);

/* Get a floating-point representation of the value stored at specified memory location */
float m4GetVal(enum m4Type type, char *posn);

/* Print the formatted value of the element of type `type' and value `val' */
void m4PrintVal(enum m4Type type, float val);

/* Load the field's value from the PSU into buf */
int m4GetConfig(struct usb_dev_handle *dev, struct m4ConfigField *field, char *buf);

/* Parse a value (123, 1.23 or 00:00:10) into buffer, encoding it */
int m4ParseValue(enum m4Type type, char const *strval, char *buf);

/* Write the human-readable value to the field on the PSU */
int m4SetConfig(struct usb_dev_handle *dev, struct m4ConfigField *field, char const *strval);

/* Print the status variables in the (raw) buffer */
void m4PrintDiag(char *buf);

/* Find the m4ConfigField of the given name */
int m4ConfigField(char const *name);

#ifdef __cplusplus
}
#endif	// __cplusplus

#endif


src/m4api.c
/*
 * m4api, a tool for M4-ATX DC-DC power supplies
 * (c) 2009-2010 Ken Tossell <ktossell@umd.edu>
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License, version 2.1,
 * as published by the Free Software Foundation. 
 *
 * This software is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this software; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#include <stdio.h>
#include <usb.h>

#include "m4atx_battery_monitor/m4api.h"

static size_t m4TypeLengths[13] = {
 1, 1, 1, 1, 1, 1, 1, 2,
 1, 2, 2, 1, 1
};

static float m4TypeConversionsV1[13] = {
 0.1123, 0.076625, 0.0375, 0.0188136,
 1.0, 1.0, 1.0, 1.0,
 10.0, 10.0, 1.0, 1.0,
 1.0
};


static float m4TypeConversionsV2[13] = {
 0.1572, 0.1165, 0.0389, 0.0195,
 1.0, 1.0, 1.0, 1.0,
 10.0, 10.0, 1.0, 1.0,
 1.0
};

static float *m4TypeConversions = NULL;

struct m4Version m4CurrentVersion = {
    .major = 0,
    .minor = 0,
};

static int m4TypeForms[13] = {
 M4_FLOAT, M4_FLOAT, M4_FLOAT, M4_FLOAT,
 M4_INTEG, M4_INTEG, M4_INTEG, M4_INTEG,
 M4_INTEG, M4_INTEG, M4_TIMER, M4_INTEG,
 M4_INTEG
};

char* m4TypeDescs[14] = {
  "V", "V", "V", "V",
  "deg. C", "sec", "ms", "ms",
  "ms", "ms", "", "tries",
  ""
};

struct m4DiagField m4DiagFields[6] = {
  {M4_VLT_12_11, 2, "VIN", "Input voltage"},
  {M4_VLT_12_11, 3, "IGN", "Ignition voltage"},
  {M4_VLT_33_01, 4, "33V", "Voltage on 3.3V rail"},
  {M4_VLT_5_03, 5, "5V", "Voltage on 5V rail"},
  {M4_VLT_12_07, 6, "12V", "Voltage on 12V rail"},
  {M4_DEG, 12, "TEMP", "Temperature"},
};

size_t m4NumDiagFields = sizeof(m4DiagFields) / sizeof(m4DiagFields[0]);

struct m4ConfigField m4ConfigFields[47] = {
  {M4_VLT_12_11, 0x00, "IGN_HIGH", "Ignition HIGH limit"},
  {M4_VLT_12_11, 0x01, "IGN_LOW", "Ignition LOW limit"},
  {M4_MSC_10_1, 0x02, "IGN_DBC", "Ignition debounce"},
  {M4_SEC, 0x03, "PSU_DELAY", "Delay before PSU startup"},
  {M4_VLT_12_11, 0x04, "VIN_MIN_START", "Min. VIN voltage at PSU startup"},
  {M4_VLT_12_11, 0x05, "VIN_MIN_ON", "Min. VIN voltage while all rails are on"},
  {M4_VLT_12_11, 0x06, "VIN_MIN_5V", "Min. VIN voltage while 5VSB is on"},
  {M4_VLT_12_11, 0x07, "VIN_MAX", "Max. allowed VIN voltage"},
  {M4_VLT_12_07, 0x08, "12V_MAX", "Max. limit for 12V"},
  {M4_VLT_12_07, 0x09, "12V_MIN", "Min. limit for 12V"},
  {M4_VLT_5_03, 0x0a, "5V_MAX", "Max. limit for 5V"},
  {M4_VLT_5_03, 0x0b, "5V_MIN", "Min. limit for 5V"},
  {M4_VLT_33_01, 0x0c, "33V_MAX", "Max. limit for 3.3V"},
  {M4_VLT_33_01, 0x0d, "33V_MIN", "Min. limit for 3.3V"},
  {M4_MSC_1_1, 0x0e, "12V_TIME", "12V rail startup time"},
  {M4_MSC_1_1, 0x0f, "33V_TIME", "3.3V rail startup time"},
  {M4_MSC_10_2, 0x10, "PWRSW", "PWRSW 'push-down' time"},
  {M4_TRY, 0x12, "PSU_ON_TIME", "Try to turn ON the PSU for N time"},
  {M4_SEC, 0x13, "ON_DELAY", "Delay between ON tries"},
  {M4_TRY, 0x14, "PSU_OFF_TIME", "Try to turn OFF the PSU for N time"},
  {M4_SEC, 0x15, "OFF_DELAY", "Delay between OFF tries"},
  {M4_SEC, 0x16, "EMG_TIME_5VSB", "Emergency shutdown timer for 5VSB"},
  {M4_SEC, 0x17, "EMG_TIMER", "Emergency shutdown timer"},
  {M4_MSC_10_1, 0x18, "PS_ON_0", "PS ON 0 threshold"},
  {M4_MSC_10_1, 0x19, "PS_ON_1", "PS ON 1 threshold"},
  {M4_MSC_10_2, 0x1a, "THUMP", "Thump timeout"},
  {M4_DEG, 0x1c, "TEMP_MAX", "Max. temperature"},
  {M4_DEG, 0x1d, "TEMP_MIN", "Min. temperature"},
  {M4_BYT, 0x1e, "EMG_OFF_MODE", "Emergency OFF mode selector"},
  {M4_MSC_10_1, 0x1f, "5V_SBY_DLY", "5V standby PSW-on delay"},
  {M4_TIM, 0x20, "OFF_DELAY_0", "Off-delay MODE 0"},
  {M4_TIM, 0x22, "OFF_HARD_0", "Hard-off MODE 0"},
  {M4_TIM, 0x24, "OFF_DELAY_1", "Off-delay MODE 1"},
  {M4_TIM, 0x26, "OFF_HARD_1", "Hard-off MODE 1"},
  {M4_TIM, 0x28, "OFF_DELAY_2", "Off-delay MODE 2"},
  {M4_TIM, 0x2a, "OFF_HARD_2", "Hard-off MODE 2"},
  {M4_TIM, 0x2c, "OFF_DELAY_3", "Off-delay MODE 3"},
  {M4_TIM, 0x2e, "OFF_HARD_3", "Hard-off MODE 3"},
  {M4_TIM, 0x30, "OFF_DELAY_4", "Off-delay MODE 4"},
  {M4_TIM, 0x32, "OFF_HARD_4", "Hard-off MODE 4"},
  {M4_TIM, 0x34, "OFF_DELAY_5", "Off-delay MODE 5"},
  {M4_TIM, 0x36, "OFF_HARD_5", "Hard-off MODE 5"},
  {M4_TIM, 0x38, "OFF_DELAY_6", "Off-delay MODE 6"},
  {M4_TIM, 0x3a, "OFF_HARD_6", "Hard-off MODE 6"},
  {M4_TIM, 0x3c, "OFF_DELAY_7", "Off-delay MODE 7"},
  {M4_TIM, 0x3e, "OFF_HARD_7", "Hard-off MODE 7"},
  /* big gap with no (known) fields */
  {M4_BYT, 0xff, "RESET", "Reset to factory defaults"},
};

#define VENDOR 0x04d8
#define PRODUCT 0xd001
#define READ_ENDPOINT 0x81
#define WRITE_ENDPOINT 0x01
#define TIMEOUT 3000


static usb_dev_handle *m4InitUSB() {
  struct usb_bus *bus;
  struct usb_device *dev;

  usb_init();

  if (usb_find_busses() < 0) {
    return NULL;
  }

  if (usb_find_devices() < 0) {
    return NULL;
  }

  bus = usb_get_busses();

  while (bus) {
    dev = bus->devices;

    while (dev) {
      if (dev->descriptor.idVendor == VENDOR &&
          dev->descriptor.idProduct == PRODUCT) {
	usb_dev_handle *handle = usb_open(dev);

	if (handle) {
#ifdef LIBUSB_HAS_DETACH_KERNEL_DRIVER_NP
          /* Linux usually claims HID devices for its usbhid driver. */
          usb_detach_kernel_driver_np(handle, 0);
#endif
	  if (usb_set_configuration(handle, 1) >= 0) {
	    if (usb_claim_interface(handle, 0) >= 0) {
	      if (usb_set_altinterface(handle, 0) < 0) {
	        usb_close(handle);
		return NULL;
              }
	    } else {
	      usb_close(handle);
	      return NULL;
	    }
	  } else {
	    usb_close(handle);
	    return NULL;
	  }

	  return handle;
	}
      }

      dev = dev->next;
    }

    bus = bus->next;
  }

  return NULL;
}

int m4Read(usb_dev_handle *dev, unsigned char *buf, unsigned int len, int timeout) {
  return usb_interrupt_read(dev, READ_ENDPOINT, (char*) buf, len, timeout);
}

int m4Write(usb_dev_handle *dev, unsigned char *buf, unsigned int len, int timeout) {
  return usb_interrupt_write(dev, WRITE_ENDPOINT, (char*) buf, len, timeout);
}

/* Gets the handle to the usb device and sets up version and conversions parameters */
usb_dev_handle *m4Init() {

  usb_dev_handle *dev = m4InitUSB();
  char buf[24];

  if (dev == NULL)
    return NULL;

  m4FetchDiag(dev, buf);

  if (m4CheckVersion(buf) < 0)
    return NULL;

  printf("Firmware version %d.%d\n", m4CurrentVersion.major, m4CurrentVersion.minor);

  if (m4CurrentVersion.major < 2)
    m4TypeConversions = m4TypeConversionsV1;
  else
    m4TypeConversions = m4TypeConversionsV2;

  return dev;
}


int m4FetchDiag (usb_dev_handle *dev, char *buf) {
  unsigned char pollCmd[] = {0x81, 0x00};
  
  if (m4Write(dev, pollCmd, 2, TIMEOUT) != 2)
    return -1;

  if (m4Read(dev, buf, 24, TIMEOUT) != 24)
    return -1;

  if (buf[0] != 0x21)
    return -1;
 
  return 0;
}

int m4CheckVersion(char *buf)
{
  m4CurrentVersion.major = ((buf[23]) >> 4) & 0x0f;
  m4CurrentVersion.minor = (buf[23]) & 0x0f;

  if (m4CurrentVersion.major == 0)
    return -1;

  return 0;
}

int m4GetDiag (usb_dev_handle *dev, struct m4Diagnostics *diag) {
  char buf[24];
  int field_id;

  if (m4FetchDiag(dev, buf) < 0)
    return -1;

  diag->vin = m4GetVal(M4_VLT_12_11, buf + 2);
  diag->vign = m4GetVal(M4_VLT_12_11, buf + 3);
  diag->v33 = m4GetVal(M4_VLT_33_01, buf + 4);
  diag->v5 = m4GetVal(M4_VLT_5_03, buf + 5);
  diag->v12 = m4GetVal(M4_VLT_12_07, buf + 6);
  diag->temp = m4GetVal(M4_DEG, buf + 12);

  return 0;
}

float m4GetVal(enum m4Type type, char *posn) {
  float val;
  short tmp_sh;
  int tmp_i;

  if (m4TypeConversions == NULL)
  {
    printf("ERROR: No type conversion values\n");
    exit(-1);
  }

  switch (m4TypeLengths[type]) {
    case 1:
      tmp_i = posn[0];
      if (type != M4_DEG) tmp_i &= 0xff;
      break;
    case 2:
      if (type == M4_DEG) {
        tmp_sh = (posn[0] << 8) | (posn[1]);
        tmp_i = (float) tmp_sh;
      } else {
        tmp_i = ((unsigned char) posn[0] << 8) | ((unsigned char) posn[1]);
      }
      break;
    default:
      printf("ERROR: typeLen(%d) == %lu!\n", type, m4TypeLengths[type]);
      exit(-1);
  }

  // printf("%x ", tmp_i);

  val = tmp_i * m4TypeConversions[type];

  return val;
}

void m4PrintVal(enum m4Type type, float val) {
  int tmp_i, sec, min, hr;

  switch (m4TypeForms[type]) {
    case M4_INTEG:
      if (type == M4_DEG)
	printf("%+d", (int) val);
      else
	printf("%d", (int) val);
      break;
    case M4_FLOAT:
      printf("%0.2f", val);
      break;
    case M4_TIMER:
      tmp_i = (int) val;

      if (tmp_i == 0xffff) {
	printf("never");
      } else {
	sec = tmp_i % 60;
	min = (tmp_i % 3600) / 60;
	hr = tmp_i / 3600;
	
	printf("%02d:%02d:%02d", hr, min, sec);
      }
      break;
    default:
      printf("ERROR: typeForm(%d) == %d!\n", type, m4TypeForms[type]);
      break;
  }
}

int m4GetConfig(usb_dev_handle *dev, struct m4ConfigField *field, char *buf) {
  unsigned char cmd[24] = {0xa4, 0xa1};

  cmd[2] = field->index;
  cmd[3] = m4TypeLengths[field->type];

  if (m4Write(dev, cmd, 24, TIMEOUT) != 24)
    return -1;

  if (m4Read(dev, buf, 24, TIMEOUT) != 24)
    return -1;

  if (buf[0] != 0x31)
    return -1;

  return 0;
}

int m4ParseValue(enum m4Type type, char const *strval, char *buf) {
  int intval;
  float fval;

  int hr, min, sec;

  switch(m4TypeForms[type]) {
  case M4_INTEG:
    intval = atoi(strval) / (int) m4TypeConversions[type];
    break;
  case M4_FLOAT:
    fval = atof(strval);
    intval = fval / m4TypeConversions[type];
    break;
  case M4_TIMER:
    if (!strcasecmp("never", strval))
      intval = 0xffff;
    else {
      if (sscanf(strval, "%d:%d:%d", &hr, &min, &sec) != 3)
	return -1;
      else
	intval = 3600 * hr + 60 * min + sec;
    }
  }

  if (m4TypeLengths[type] == 2) {
    buf[0] = intval >> 8;
    buf[1] = intval;
  } else {
    buf[0] = intval;
    buf[1] = 0;
  }

  return 0;
}

int m4GetFloat(usb_dev_handle *dev, enum m4FieldID fid, float *out) {
  char buf[24];
  struct m4ConfigField *field;
  
  field = &m4ConfigFields[fid];

  if (m4GetConfig(dev, field, buf))
    return -1;

  *out = m4GetVal(field->type, &buf[4]);

  return 0;
}

int m4SetFloat(usb_dev_handle *dev, enum m4FieldID fid, float val) {
  char binary[2];
  int ival;
  struct m4ConfigField *field;
  
  field = &m4ConfigFields[fid];

  ival = val / m4TypeConversions[field->type];

  if (m4TypeLengths[field->type] == 2) {
    binary[0] = ival >> 8;
    binary[1] = ival;
  } else {
    binary[0] = ival;
    binary[1] = 0;
  }

  return m4SetBinary(dev, field, binary);
}

int m4GetInteger(usb_dev_handle *dev, enum m4FieldID fid, int *out) {
  char buf[24];
  struct m4ConfigField *field;
    
  field = &m4ConfigFields[fid];

  if (m4GetConfig(dev, field, buf))
    return -1;

  *out = (int) m4GetVal(field->type, &buf[4]);

  return 0;
}

int m4SetInteger(usb_dev_handle *dev, enum m4FieldID fid, int val) {
  char binary[2];
  struct m4ConfigField *field;

  field = &m4ConfigFields[fid];

  val /= m4TypeConversions[field->type];

  if (m4TypeLengths[field->type] == 2) {
    binary[0] = val >> 8;
    binary[1] = val;
  } else {
    binary[0] = val;
    binary[1] = 0;
  }

  return m4SetBinary(dev, field, binary);
}

int m4SetConfig(usb_dev_handle *dev, struct m4ConfigField *field, char const *strval) {
  char binary[] = {0, 0};

  if (m4ParseValue(field->type, strval, binary) < 0)
    return -1;

  return m4SetBinary(dev, field, binary);
}

int m4SetBinary(usb_dev_handle *dev, struct m4ConfigField *field, char const *val) {
  char buf[24];
  unsigned char cmd[24] = {0xa4, 0xa0};

  cmd[2] = field->index;
  cmd[3] = m4TypeLengths[field->type];
  cmd[4] = val[0];
  cmd[5] = val[1];

  if (m4Write(dev, cmd, 24, TIMEOUT) != 24)
    return -1;

  if (m4Read(dev, buf, 24, TIMEOUT) != 24)
    return -1;

  if (buf[0] != 0x31 || buf[2] != cmd[2]
      || buf[4] != cmd[4] || buf[5] != cmd[5])
    return -1;

  return 0;
}

void m4PrintDiag(char *buf) {
  int field_id;
  size_t config_offset = 0;
  float value;

  for (field_id = 0; field_id < m4NumDiagFields; ++field_id) {
    printf("%s:\t", m4DiagFields[field_id].name);
    value = m4GetVal(m4DiagFields[field_id].type, buf + m4DiagFields[field_id].index);
    m4PrintVal(m4DiagFields[field_id].type, value);
    puts("");
  }
}

int m4ConfigField(char const *name) {
  int field_id;

  for (field_id = 0; field_id < M4_NUM_CONFIG_FIELDS; ++field_id) {
    if (!strcasecmp(m4ConfigFields[field_id].name, name))
      return field_id;
  }

  return -1;
}

